### PR TITLE
[CURATOR-217] Use new Watcher Removal APIs in Curator Recipes

### PR DESCRIPTION
**Pull request by Randgalt**

### CODE REVIEW

File: curator-framework/src/main/java/org/apache/curator/framework/CuratorFramework.java [hunk 1]
```diff
@@ -235,4 +235,6 @@
      * @throws InterruptedException If interrupted while waiting
      */
     public void blockUntilConnected() throws InterruptedException;
+
+    WatcherRemoveCuratorFramework newWatcherRemoveCuratorFramework();
```
Comment (cammckenzie) [inline curator-framework/src/main/java/org/apache/curator/framework/CuratorFramework.java]: Javadoc?

File: curator-framework/src/main/java/org/apache/curator/framework/WatcherRemoveCuratorFramework.java [hunk 2]
```diff
@@ -0,0 +1,24 @@
...
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package org.apache.curator.framework;
+
+public interface WatcherRemoveCuratorFramework extends CuratorFramework
+{
+    void removeWatchers();
```
Comment (cammckenzie) [inline curator-framework/src/main/java/org/apache/curator/framework/WatcherRemoveCuratorFramework.java]: Javadoc?

Comment (Randgalt) [inline curator-framework/src/main/java/org/apache/curator/framework/WatcherRemoveCuratorFramework.java]: Yeah - missed that

File: curator-framework/src/main/java/org/apache/curator/framework/imps/WatcherRemovalManager.java [hunk 3]
```diff
@@ -0,0 +1,138 @@
...
+            localEntries = Sets.newHashSet(entries);
+        }
+        for ( WrappedWatcher entry : localEntries )
+        {
+            try
+            {
+                log.debug("Removing watcher for path: " + entry.path);
+                RemoveWatchesBuilderImpl builder = new RemoveWatchesBuilderImpl(client);
+                builder.prepInternalRemoval(entry);
+                builder.pathInForeground(entry.path);
```
Comment (Randgalt) [inline curator-framework/src/main/java/org/apache/curator/framework/imps/WatcherRemovalManager.java]: I wonder if this should be a background/async call instead

File: curator-framework/src/test/java/org/apache/curator/framework/imps/TestWatcherRemovalManager.java [hunk 4]
```diff
@@ -0,0 +1,251 @@
...
+            }
+        };
+        removerClient.checkExists().usingWatcher(watcher).forPath("/hey");
+
+        List<String> existWatches = WatchersDebug.getExistWatches(client.getZookeeperClient().getZooKeeper());
+        Assert.assertEquals(existWatches.size(), 1);
+
+        removerClient.removeWatchers();
+
+        new Timing().sleepABit();
```
Comment (cammckenzie) [inline curator-framework/src/test/java/org/apache/curator/framework/imps/TestWatcherRemovalManager.java]: Could this test be more reliable with a latch in the process() method of the watcher? If for whatever reason the remove takes longer than Timing().sleepABit() then the tests using this method will currently fail.

File: curator-framework/src/test/java/org/apache/curator/framework/imps/TestWatcherRemovalManager.java [hunk 5]
```diff
@@ -0,0 +1,251 @@
...
+                        latch.countDown();
+                    }
+                }
+            };
+
+            removerClient.checkExists().usingWatcher(watcher).forPath("/yo");
+            Assert.assertEquals(removerClient.getRemovalManager().getEntries().size(), 1);
+            removerClient.create().forPath("/yo");
+
+            timing.sleepABit();
```
Comment (cammckenzie) [inline curator-framework/src/test/java/org/apache/curator/framework/imps/TestWatcherRemovalManager.java]: Latch after reset to ensure this is deterministic?

File: curator-framework/src/main/java/org/apache/curator/framework/imps/WatcherRemovalManager.java (hunk 3 above)
Comment (cammckenzie) [inline curator-framework/src/main/java/org/apache/curator/framework/imps/WatcherRemovalManager.java]: That would probably be a good idea. You could do it as a guaranteed background operation.

**Comment by cammckenzie** [comment]
Other than my minor comments, this looks good to me. Still needs to have the updates to CURATOR-161 merged into it also as currently one of the CURATOR-161 related unit tests is breaking.

Are there any other recipes that require watcher removal?

File: curator-framework/src/test/java/org/apache/curator/framework/imps/TestWatcherRemovalManager.java (hunk 4 above)
Comment (Randgalt) [inline curator-framework/src/test/java/org/apache/curator/framework/imps/TestWatcherRemovalManager.java]: done

File: curator-framework/src/test/java/org/apache/curator/framework/imps/TestWatcherRemovalManager.java (hunk 5 above)
Comment (Randgalt) [inline curator-framework/src/test/java/org/apache/curator/framework/imps/TestWatcherRemovalManager.java]: done

**Comment by Randgalt** [comment]
All comments applied, latest CURATOR-161 merged.

**Comment by cammckenzie** [comment]
LGTM

**Pull request closed without merging**